### ISSUE TITLE
Substorage: Different "views" on storage

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/openpathsampling#330]
[This may already be a part of storage (or the new storage); if so, we can close. If it isn't already implemented, let's make a new PR of it (don't add any more features to #300).]

Pretty frequently, we think about just using a subset of the total steps from a calculation. For example, if I'm using block averaging to get estimates for the errors, I may want to split the steps into 10-20 subgroups, and do the analysis on each of those. Or I might want to compare a long (10+k steps) run of a MSTIS run with a shorter SRTIS which gets its fixed bias from the first 1k steps of the MSTIS.

It would be nice to have a simple notation for this. At the very least, I think we could do something like:

```python
substore = storage.substorage(step_range=slice(0,2000,2))
```

and it might even be possible to do

```python
substore = storage[slice(0,2000,2)]
```

Importantly, such a substorage should be **read-only**. Otherwise there's ambiguity about what should be saved. It should also still have access to all the ensembles, etc. in the storage. In my mind, the only thing that changes is how the iterator over steps works. In the examples above,  I would like for:

```python
for step in substore.steps:
    analyze(step)
```

to be the same as

```python
for step in storage.steps[slice(0,2000,2)]:
    analyze(step)
```

Another option might be to change the iterator so that `for step in storage` is the same as `for step in storage.steps`; I think that's going to be the most common usage, but I don't know if that's necessarily safe to do.

It might also be worth thinking about the possibility of allowing ensembles/lists of ensembles or replica IDs/lists of replica IDs to define substorages as well. Again, this would be a convenience for analysis, which often wants to track one replica ID (for replica exchange) or one ensemble (for ensemble property analysis). We have ways to do that already, but this might be cleaner and more user-friendly. This might be implemented by having:

```python
# in Storage
def substorage(self, step_range=None, ensembles=None, replicas=None):
    """If more than one argument is defined, the combination is treated as an 'and'."""
```

**Comment by jhprinz** [comment]
Alright. This is very neat. And the first where only `mcsteps` is affected is almost trivial. I can make a kind of proxy that gives you all stores and override the `mcsteps` in the proxy. As long as no other stored object will reference any mcsteps this is not a problem. Otherwise a stored object will call the altered store to get the objects and use wrong indices.

So, if you do that then effectively `st.steps[10]` will call `st.steps[20]` 

If you do the same for ensembles this will be a problem. 

The easiest way to do what you want is to not rely on storage but on the store that you get. Remember that I call the whole thing `storage` and the variable lists for single objects `store`. And a store is just a list. So you would call the anaylsis with
```
results = analysis(storage.steps[::2], storage.ensembles['ensA', 'ensB']):
def analysis(steps, ensembles):
    for step in steps:
```
instead of
```
results = analysis(storage)
def analysis(storage):
    for step in storage.steps:
```

I still like the idea of a view on storage (like in numpy). Could you elaborate where the limit to a few ensembles is used? You will surely not loop over all ensembles, but specify a few ones, right?

**Comment by dwhswenson** [comment]
> Could you elaborate where the limit to a few ensembles is used?

I was thinking about trying to return only the samples representing certain ensembles. For example, when you want to calculate histograms to show ensemble properties (maximum lambda, path length) you only need one ensemble at a time.

But as I think more, my guess is that the overhead from creating new `SampleSet`s wouldn't be worth it, and I'd rather keep the return type the same. And really, it isn't so hard to type `step.active[ensemble]` or `step.active[replica]`, so nevermind on that.

> The easiest way to do what you want is to not rely on storage but on the store that you get.

Yes, but the analysis functions call other analysis functions. What I'm looking for is a way to avoid making the function signatures huge for everything. I don't want to write functions that look like:

```python
run_analysis(storage.steps[0:2000:2], storage.ensembles, storage.collectivevariables, storage.networks)
```

when it could just be

```python
substorage = storage.substorage(slice(0,2000,2))
run_analysis(substorage)
```

**Comment by jhprinz** [comment]
I talked with @dwhswenson about this and we thought it would be best to create a substorage that will iterate over a certain subset instead of all steps.

The advantange is that we can keep all ids the way they are. Only if you use
```
for step in storage.steps:
   # do something
```
in your code it will loop over the subset. This should make all analysis work only on parts of your data.
To avoid problems of having missing intermediates, etc we should not allow to use slicing with steps other than `1`.

**Comment by jhprinz** [comment]
As mentioned in #191. We can use this to do the analysis on sub-blocks of steps. This step should be easy once the block-selection works.

**Comment by jhprinz** [comment]
Should be closed by #395 already. Closing.
